File: client/ts/src/connection/checker.ts
// Copyright 2026 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

import { type UnaryClient } from "@synnaxlabs/freighter";
import {
  ClockSkewCalculator,
  type CrudeTimeSpan,
  errors,
  migrate,
  TimeSpan,
  TimeStamp,
} from "@synnaxlabs/x";
import { z } from "zod";

export const statusZ = z.enum(["disconnected", "connecting", "connected", "failed"]);
export type Status = z.infer<typeof statusZ>;

export const stateZ = z.object({
  status: statusZ,
  error: z.instanceof(Error).optional(),
  message: z.string().optional(),
  clusterKey: z.string(),
  clientVersion: z.string(),
  clientServerCompatible: z.boolean(),
  nodeVersion: z.string().optional(),
  clockSkew: TimeSpan.z.default(TimeSpan.ZERO),
  clockSkewExceeded: z.boolean().default(false),
});
export interface State extends z.infer<typeof stateZ> {}

const responseZ = z.object({
  clusterKey: z.string(),
  nodeVersion: z.string().optional(),
  nodeTime: TimeStamp.z,
});
const requestZ = z.void();

const DEFAULT: State = {
  clusterKey: "",
  status: "disconnected",
  error: undefined,
  message: "Disconnected",
  clientServerCompatible: false,
  clientVersion: __VERSION__,
  clockSkew: TimeSpan.ZERO,
  clockSkewExceeded: false,
};

const createWarning = (
  nodeVersion: string | null,
  clientVersion: string,
  clientIsNewer: boolean,
): string => {
  const toUpgrade = clientIsNewer ? "Core" : "client";
  return `The Synnax Core version ${nodeVersion != null ? `${nodeVersion} ` : ""}is too ${clientIsNewer ? "old" : "new"} for client version ${clientVersion}.
  This may cause compatibility issues. We recommend updating the ${toUpgrade}. For more information, see
  https://docs.synnaxlabs.com/reference/client/resources/troubleshooting#old-${toUpgrade.toLowerCase()}-version`;
};

/** Polls a synnax cluster for connectivity information. */
export class Checker {
  static readonly DEFAULT: State = DEFAULT;
  private readonly liveState: State;
  private readonly pollFrequency: TimeSpan;
  private readonly client: UnaryClient;
  private readonly name?: string;
  private interval?: NodeJS.Timeout;
  private readonly clientVersion: string;
  private readonly onChangeHandlers: Array<(state: State) => void> = [];
  static readonly connectionStateZ = stateZ;
  private versionWarned = false;
  private readonly skewCalc: ClockSkewCalculator;
  private readonly clockSkewThreshold: TimeSpan;
  private checking = false;

  /**
   * @param client - The transport client to use for connectivity checks.
   * @param pollFreq - The frequency at which to poll the cluster for
   *   connectivity information.
   */
  constructor(
    client: UnaryClient,
    pollFreq: CrudeTimeSpan = TimeSpan.seconds(30),
    clientVersion: string,
    name?: string,
    clockSkewThreshold: CrudeTimeSpan = TimeSpan.seconds(1),
  ) {
    this.liveState = { ...DEFAULT };
    this.client = client;
    this.pollFrequency = new TimeSpan(pollFreq);
    this.clientVersion = clientVersion;
    this.name = name;
    this.skewCalc = new ClockSkewCalculator();
    this.clockSkewThreshold = new TimeSpan(clockSkewThreshold).abs();
    void this.check();
    this.start();
  }

  /** Stops the connectivity client from polling the cluster for connectivity */
  stop(): void {
    if (this.interval != null) clearInterval(this.interval);
  }

  /**
   * Executes a connectivity check and updates the client status and error, as
   * well as calling any registered change handlers.
   */
  async check(): Promise<State> {
    const prevStatus = this.liveState.status;
    const prevSkewExceeded = this.liveState.clockSkewExceeded;
    const measureSkew = !this.checking;
    this.checking = true;
    try {
      if (measureSkew) this.skewCalc.start();
      const res = await this.client.send(
        "/connectivity/check",
        undefined,
        requestZ,
        responseZ,
      );
      if (measureSkew) {
        this.skewCalc.end(res.nodeTime);
        this.liveState.clockSkew = this.skewCalc.skew;
        this.liveState.clockSkewExceeded = this.skewCalc.exceeds(
          this.clockSkewThreshold,
        );
        if (this.liveState.clockSkewExceeded) {
          const direction = this.skewCalc.skew.valueOf() > 0n ? "ahead of" : "behind";
          console.warn(
            `Measured excessive clock skew between this host and ` +
              `the Synnax Core. This host is ${direction} the Synnax Core ` +
              `by approximately ${this.skewCalc.skew.abs().toString()}.`,
          );
        }
      }
      const nodeVersion = res.nodeVersion;
      const { clientVersion } = this;
      if (nodeVersion == null) {
        this.liveState.clientServerCompatible = false;
        if (!this.versionWarned) {
          console.warn(createWarning(null, clientVersion, true));
          this.versionWarned = true;
        }
      } else if (
        !migrate.versionsEqual(clientVersion, nodeVersion, {
          checkMajor: true,
          checkMinor: true,
          checkPatch: false,
        })
      ) {
        this.liveState.clientServerCompatible = false;
        if (!this.versionWarned) {
          console.warn(
            createWarning(
              nodeVersion,
              clientVersion,
              migrate.semVerNewer(clientVersion, nodeVersion),
            ),
          );
          this.versionWarned = true;
        }
      } else this.liveState.clientServerCompatible = true;
      this.liveState.status = "connected";
      this.liveState.message = `Connected to ${this.name ?? "cluster"}`;
      this.liveState.clusterKey = res.clusterKey;
      this.liveState.nodeVersion = res.nodeVersion;
      this.liveState.clientVersion = this.clientVersion;
    } catch (err) {
      this.liveState.status = "failed";
      this.liveState.error = errors.fromUnknown(err);
      this.liveState.message = this.state.error?.message;
    } finally {
      this.checking = false;
    }
    const changed =
      prevStatus !== this.liveState.status ||
      prevSkewExceeded !== this.liveState.clockSkewExceeded;
    if (this.onChangeHandlers.length > 0 && changed)
      this.onChangeHandlers.forEach((handler) => handler(this.state));
    return this.state;
  }

  /** @returns a copy of the current client state. */
  get state(): State {
    return { ...this.liveState };
  }

  /** @param callback - The function to call when the client status changes. */
  onChange(callback: (state: State) => void): void {
    this.onChangeHandlers.push(callback);
  }

  private start(): void {
    this.interval = setInterval(() => {
      void this.check();
    }, this.pollFrequency.milliseconds);
  }
}

File: client/cpp/connection/checker.cpp
// Copyright 2026 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

#include <chrono>
#include <sstream>

#include "client/cpp/connection/checker.h"

namespace synnax::connection {
const std::string TROUBLESHOOTING_URL =
    "https://docs.synnaxlabs.com/reference/client/resources/troubleshooting";

std::pair<int, int> parse_version(const std::string &v) {
    std::istringstream ss(v);
    int major = 0, minor = 0;
    char dot;
    ss >> major >> dot >> minor;
    return {major, minor};
}

bool versions_compatible(const std::string &v1, const std::string &v2) {
    return parse_version(v1) == parse_version(v2);
}

bool client_is_newer(const std::string &client_ver, const std::string &node_ver) {
    return parse_version(client_ver) > parse_version(node_ver);
}

std::string create_version_warning(
    const std::string &node_version,
    const std::string &client_version,
    const bool is_client_newer
) {
    const auto to_upgrade = is_client_newer ? "Core" : "client";
    const auto age = is_client_newer ? "old" : "new";
    std::ostringstream ss;
    ss << "The Synnax core version ";
    if (!node_version.empty()) ss << node_version << " ";
    ss << "is too " << age << " for client version " << client_version
       << ". This may cause compatibility issues. We recommend updating the "
       << to_upgrade << ". For more information, see " << TROUBLESHOOTING_URL << "#old-"
       << to_upgrade << "-version";
    return ss.str();
}

Checker::Checker(
    std::unique_ptr<CheckClient> client,
    const x::telem::TimeSpan poll_freq,
    std::string client_version,
    std::string name,
    const x::telem::TimeSpan clock_skew_threshold
):
    client(std::move(client)),
    poll_freq(poll_freq),
    client_version(std::move(client_version)),
    name(std::move(name)),
    clock_skew_threshold(clock_skew_threshold) {
    this->_state.client_version = this->client_version;
    this->running = true;
    this->check();
    this->poll_thread = std::thread([this] { this->run(); });
}

Checker::~Checker() {
    this->stop();
}

void Checker::stop() {
    {
        std::lock_guard lk(this->cv_mu);
        this->running = false;
    }
    this->cv.notify_all();
    if (this->poll_thread.joinable()) this->poll_thread.join();
}

State Checker::check() {
    Status prev_status;
    bool prev_skew_exceeded;
    {
        std::lock_guard lk(this->mu);
        prev_status = this->_state.status;
        prev_skew_exceeded = this->_state.clock_skew_exceeded;
    }

    this->skew_calc.start();
    google::protobuf::Empty req;
    auto [res, err] = this->client->send("/connectivity/check", req);

    State snapshot;
    {
        std::lock_guard lk(this->mu);
        if (err) {
            this->_state.status = Status::FAILED;
            this->_state.error = err;
            this->_state.message = err.message();
        } else {
            this->skew_calc.end(x::telem::TimeStamp(res.node_time()));
            this->_state.clock_skew = this->skew_calc.skew();
            this->_state.clock_skew_exceeded = this->skew_calc.exceeds(
                this->clock_skew_threshold
            );
            if (this->_state.clock_skew_exceeded) {
                auto direction = this->skew_calc.skew() > x::telem::TimeSpan::ZERO()
                                   ? "ahead of"
                                   : "behind";
                LOG(WARNING) << "Measured excessive clock skew between this host and "
                                "the Synnax core. This host is "
                             << direction << " the Synnax core by approximately "
                             << this->skew_calc.skew().abs() << ".";
            }
            const auto &nv = res.node_version();
            if (nv.empty()) {
                this->_state.client_server_compatible = false;
                if (!this->version_warned) {
                    LOG(WARNING)
                        << create_version_warning("", this->client_version, true);
                    this->version_warned = true;
                }
            } else if (!versions_compatible(this->client_version, nv)) {
                this->_state.client_server_compatible = false;
                if (!this->version_warned) {
                    LOG(WARNING) << create_version_warning(
                        nv,
                        this->client_version,
                        client_is_newer(this->client_version, nv)
                    );
                    this->version_warned = true;
                }
            } else {
                this->_state.client_server_compatible = true;
            }
            this->_state.status = Status::CONNECTED;
            this->_state.message = "Connected to " +
                                   (this->name.empty() ? "cluster" : this->name);
            this->_state.cluster_key = res.cluster_key();
            this->_state.node_version = nv;
        }
        snapshot = this->_state;
    }

    const bool changed = prev_status != snapshot.status ||
                         prev_skew_exceeded != snapshot.clock_skew_exceeded;
    if (changed) this->fire_handlers(snapshot);
    return snapshot;
}

State Checker::state() const {
    std::lock_guard lk(this->mu);
    return this->_state;
}

void Checker::on_change(std::function<void(const State &)> handler) {
    this->on_change_handlers.push_back(std::move(handler));
}

void Checker::run() {
    while (true) {
        std::unique_lock lk(this->cv_mu);
        const auto nanos = this->poll_freq.nanoseconds();
        if (this->cv.wait_for(lk, std::chrono::nanoseconds(nanos), [this] {
                return !this->running.load();
            }))
            return;
        lk.unlock();
        this->check();
    }
}

void Checker::fire_handlers(const State &s) {
    for (const auto &handler: this->on_change_handlers)
        handler(s);
}
}
